File: Behavioral/COR/DIY.java
package COR;

public class DIY {
    public static void main(String[] args) {
        var receipt = new Receipt(null);
        var production = new Production(receipt);
        var supply = new Supply(production,10);
        var invoice = new Invoice(supply);

        var delivery = new DeliveryProcess(invoice);
        delivery.handle(new Order("Soup Can",2));
        delivery.handle(new Order("Soap Bar",2));
        delivery.handle(new Order("Soap Bar",10));

    }
    
}

class DeliveryProcess{
    private Handler handler;

    public DeliveryProcess(Handler handler){
        this.handler = handler;
    }
        public void handle(Order order){
        handler.handle(order);
        }
    }



class Order{
    private String prodName;
    private int quantity;

    public Order(String prodName, int quantity){
        this.prodName = prodName;
        this.quantity = quantity;
    }

    public String getProdname(){
        return prodName;
    }

    public int getQuantity(){
        return quantity;
    }
}

class Receipt extends Handler{

    public Receipt(Handler nextProcess) {
            super(nextProcess);
        }
    
        @Override
    boolean doHandle(Order order) {
        System.out.println("Receipt: Printing Receipt...");
        System.out.println("Receipt done. Delivery Process executed properly.");
       return false;
    }
    

}
class Production extends Handler{

    public Production(Handler nextProcess) {
            super(nextProcess);
            
        }
    
        @Override
    boolean doHandle(Order order) {
        System.out.println("Production: Producing " + order.getProdname() + "...");
        System.out.println("Item added for Deliver.");
        return false;

    }

}

class Supply extends Handler{
    private int stock;

    public Supply(Handler nextProcess, int stock) {
            super(nextProcess);
            this.stock=stock;
            }
    
        @Override
    boolean doHandle(Order order) {
     System.out.println("Supply: Supplier checking stocks...");
     if(order.getQuantity()> stock){
        System.out.println("Supply: No stocks available.");
        return true;
     }
     stock -= order.getQuantity();
     System.out.println("Supply: Stocks available. Remaining stock: " + stock + ".");
    return false;
    }

}

class Invoice extends Handler{

    public Invoice(Handler nextProcess) {
            super(nextProcess);
        }
    
        @Override
    boolean doHandle(Order order) {
       System.out.println("Invoice: Preparing for invoice of " + order.getProdname() + "..." );
       System.out.println("Invoice Sent.");


       return false;
    }

}

abstract class Handler{
    private Handler nextProcess;
    public Handler(Handler nextProcess){
        this.nextProcess = nextProcess;
    }
    public void handle(Order order){
        if(doHandle(order)){
            return;
        }
        if(nextProcess !=null){
            nextProcess.handle(order);
        }
    }

    abstract boolean doHandle(Order order);
}

File: TS/Behavioral/COR/DIY.ts
class Order {
  private prodName: string;
  private quantity: number;

  constructor(prodName: string, quantity: number) {
    this.prodName = prodName;
    this.quantity = quantity;
  }

  getProdName(): string {
    return this.prodName;
  }

  getQuantity(): number {
    return this.quantity;
  }
}

abstract class Handler {
  private nextProcess: Handler | undefined;

  constructor(nextProcess?: Handler) {
    this.nextProcess = nextProcess;
  }

  handle(order: Order): void {
    if (this.doHandle(order)) {
      return;
    }
    if (this.nextProcess != null) {
      this.nextProcess.handle(order);
    }
  }

  abstract doHandle(order: Order): boolean;
}

class Receipt extends Handler {
  constructor(nextProcess?: Handler) {
    super(nextProcess);
  }

  doHandle(order: Order): boolean {
    console.log("Receipt: Printing Receipt...");
    console.log("Receipt done. Delivery Process executed properly.");
    return false;
  }
}

class Production extends Handler {
  constructor(nextProcess: Handler) {
    super(nextProcess);
  }

  doHandle(order: Order): boolean {
    console.log("Production: Producing " + order.getProdName() + "...");
    console.log("Item added for Deliver.");
    return false;
  }
}

class Supply extends Handler {
  private stock: number;

  constructor(nextProcess: Handler, stock: number) {
    super(nextProcess);
    this.stock = stock;
  }

  doHandle(order: Order): boolean {
    console.log("Supply: Supplier checking stocks...");
    if (order.getQuantity() > this.stock) {
      console.log("Supply: No stocks available.");
      return true;
    }
    this.stock -= order.getQuantity();
    console.log(
      "Supply: Stocks available. Remaining stock: " + this.stock + "."
    );
    return false;
  }
}

class Invoice extends Handler {
  constructor(nextProcess: Handler) {
    super(nextProcess);
  }

  doHandle(order: Order): boolean {
    console.log(
      "Invoice: Preparing for invoice of " + order.getProdName() + "..."
    );
    console.log("Invoice Sent.");
    return false;
  }
}

class DeliveryProcess {
  private handler: Handler;

  constructor(handler: Handler) {
    this.handler = handler;
  }

  handle(order: Order): void {
    this.handler.handle(order);
  }
}

// The Client
const receipt = new Receipt();
const production = new Production(receipt);
const supply = new Supply(production, 10);
const invoice = new Invoice(supply);

const delivery = new DeliveryProcess(invoice);
delivery.handle(new Order("Soup Can", 2));
delivery.handle(new Order("Soap Bar", 2));
delivery.handle(new Order("Soap Bar", 10));
